ekncontent/ekncontent/eknc-engine.c

```diff
@@ -0,0 +1,499 @@
+{
+  EkncDomain *domain = EKNC_DOMAIN (source);
+  g_autoptr(GTask) task = user_data;
+  GError *error = NULL;
+
+  EkncContentObjectModel *model;
+  if (!(model = eknc_domain_get_object_finish (domain, result, &error)))
+    {
+      g_task_return_error (task, error);
+      return;
+    }
+
+  g_task_return_pointer (task, model, g_object_unref);
+}
+
+/**
+ * eknc_engine_get_object_for_app:
+ * @self: the engine
+ * @id: the ekn id of the object to load
+ * @app_id: the id of the application to load the object from
+ * @cancellable: (allow-none): optional #GCancellable object, %NULL to ignore.
+ * @callback: (scope async): callback to call when the request is satisfied.
+ * @user_data: (closure): the data to pass to callback function.
+ *
+ * Asynchronously load an object model for the given ekn_id
+ */
+void
+eknc_engine_get_object_for_app (EkncEngine *self,
+                                const gchar *id,
+                                const gchar *app_id,
+                                GCancellable *cancellable,
+                                GAsyncReadyCallback callback,
+                                gpointer user_data)
+{
+  g_return_if_fail (EKNC_IS_ENGINE (self));
+  g_return_if_fail (id != NULL);
+  g_return_if_fail (G_IS_CANCELLABLE (cancellable) || cancellable == NULL);
+  g_return_if_fail (app_id && *app_id);
+
+  g_autoptr(GTask) task = g_task_new (self, cancellable, callback, user_data);
+  GError *error;
+  EkncDomain *domain;
+  if (!(domain = eknc_engine_get_domain_for_app (self, app_id, &error)))
+    {
+      g_task_return_error (task, error);
+      return;
+    }
+
+  eknc_domain_get_object (domain, id, cancellable, on_domain_object_finished, g_steal_pointer (&task));
+}
+
+/**
+ * eknc_engine_get_object_for_app_finish:
+ * @self: the engine
+ * @result: the #GAsyncResult that was provided to the callback.
+ * @error: #GError for error reporting.
+ *
+ * Finish a eknc_engine_get_object_for_app call.
+ *
+ * Returns: (transfer full): the content object model that was fetched
+ */
+EkncContentObjectModel *
+eknc_engine_get_object_for_app_finish (EkncEngine *self,
+                                       GAsyncResult *result,
+                                       GError **error)
+{
+  g_return_val_if_fail (EKNC_IS_ENGINE (self), NULL);
+  g_return_val_if_fail (G_IS_ASYNC_RESULT (result), NULL);
+
+  return g_task_propagate_pointer (G_TASK (result), error);
+}
+
+static void
+on_domain_query_finished (GObject *source,
+                          GAsyncResult *result,
+                          gpointer user_data)
+{
+  EkncDomain *domain = EKNC_DOMAIN (source);
+  g_autoptr(GTask) task = user_data;
+  GError *error = NULL;
+
+  EkncQueryResults *results;
+  if (!(results = eknc_domain_query_finish (domain, result, &error)))
+    {
+      g_task_return_error (task, error);
+      return;
+    }
+
+  g_task_return_pointer (task, results, g_object_unref);
+}
+
+static void
+on_domain_fix_query_finished (GObject *source,
```

Comment:
nitpick: Seems like the name of this callback ought to indicate that it's part of a `eknc_engine_query()` operation, because it continues with `eknc_domain_query()` at the end.